### PR TITLE
Add test against directory

### LINKED ISSUE #85: execute tests against a directory
it would be nice if tests could be executed against a complete directory of files

### PR DESCRIPTION
**Pull request by dylanhitt**
closes #85. I thought I would take a stab at this. I am mainly using this as a learning experience and would love feedback. I thought this would be a good issue to learn the codebase.

This change will allow a user to execute`./commander test examples/` or `./commander test --file-order examples/`. Here are the following outputs respectively: 

```sh
dylanhitt@Dylans-MacBook-Pro commander (add-test-against-directory) $ ./commander test examples/ 
Starting test against directory: examples/...

✓ [minimal_test.yaml] [local] echo hello
✓ [environment.yaml] [local] echo $KEY
✓ [commander.yaml] [local] echo hello
✓ [environment.yaml] [local] it should overwrite env
✗ [commander.yaml] [local] it should fail
✓ [commander.yaml] [local] it should print hello world
✓ [commander.yaml] [local] it should validate exit code

Results

✗ [commander.yaml] [local] 'it should fail', on property 'Stderr'

Expected

/bin/sh: invalid: command not found

to contain

/bin/sh: 1: invalid: not found


Duration: 0.041s
Count: 7, Failed: 1

Test suite failed, use --verbose for more detailed output
``` 
```sh
dylanhitt@Dylans-MacBook-Pro commander (add-test-against-directory) $ ./commander test --file-order examples/ 
Starting test against directory: examples/...

✓ [commander.yaml] [local] it should print hello world
✓ [commander.yaml] [local] it should validate exit code
✓ [commander.yaml] [local] echo hello
✗ [commander.yaml] [local] it should fail
✓ [environment.yaml] [local] echo $KEY
✓ [environment.yaml] [local] it should overwrite env
✓ [minimal_test.yaml] [local] echo hello

Results

✗ [commander.yaml] [local] 'it should fail', on property 'Stderr'

Expected

/bin/sh: invalid: command not found

to contain

/bin/sh: 1: invalid: not found


Duration: 0.036s
Count: 7, Failed: 1

Test suite failed, use --verbose for more detailed output
```

The change also supports testing multiple files for a test that has the same title. Here is an example:

```sh
dylanhitt@Dylans-MacBook-Pro commander (add-test-against-directory) $ ./commander test examples/ "echo hello" 
Starting test against directory: examples/...

✓ [commander.yaml] [local] echo hello
✓ [minimal_test.yaml] [local] echo hello

Duration: 0.012s
Count: 2, Failed: 0

File Errors:
[environment.yaml] could not find test echo hello!
```

If it's something you would consider merging let me know and I can finish up writing the unit test and integration tests for the change as well as the docs. Just opening this now to get the ball rolling (:

Thanks!

## Checklist

 - [x] Added unit / integration tests for windows, macOS and Linux? 
 - [x] Added a changelog entry in [CHANGELOG.md](../CHANGELOG.md)?
 - [x] Updated the documentation ([README.md](../README.md), [docs](../docs))?
 - [x] Does your change work on `Linux`, `Windows` and `macOS`?

### CODE REVIEW

**Comment by SimonBaeumer** [comment]
Hey @dylanhitt,
thanks and sorry for the late answer! I will take a look at it today :)
Best regards!

**Comment by dylanhitt** [comment]
Hey @SimonBaeumer,
I actually did quite a bit of refactoring the last two days. Here's a new commit for ya.

**Comment by SimonBaeumer** [comment]
First of all your solution works and I like the feature already :D
There are some things in the pull request I want to tackle and diving a little bit deeper into it why the implemented solution would increase the technical debt. 

- Adding `--dir` flag
I would prefer a `--dir` flag because than I intentionally said that a directory of suites should be executed. Without the `--dir` it would not obvious if I started tests accidentally against a directory and just don't get any output.
Think of yourself trying to use the tool and suddenly you forgot to provide the filename and you don't get an error message - that can be really annoying "errors" :D

- Removing `--file-order` flag / synchronous execution
The `--file-order` flag was at first a little bit irritating to me because it just sorts the results alphabetically so they will be grouped by file name if printed to the console.
A huge draw back on your solution is that in [cli.go:56](https://github.com/SimonBaeumer/commander/pull/120/files#diff-7e41fcd909802179700c23b43db33e32R56) you basically will not write any results until the results channel is closed (the channel is closed when the `for` loop exited). 
I would remove the `--file-order` flag and execute the tests synchronously and guarantee the alphabetical order by default.
A `--concurrent=N` flag could enable concurrent execution of test files where `N` is the number of `goroutines`. At the moment this feature would be too much.

- Unwrap `runtime.TestResult`
This is a little bit more complex as it involves different components of the system. 
Commander handles every file as a separate test suite so it is totally okay to iterate over all files inside the directory and just executing them.
A problem occurs because the `cli.go` requires a `runtime.TestResult` channel for the output which is not really nice.
I try to explain inside the source code where you can read it up more clearly.

- Points to think of
  - Exiting immediately if a file could not be read
  - What about subdirs?

File: pkg/app/test_command.go [hunk 1]
```diff
@@ -2,53 +2,112 @@ package app
...
+	return nil
+}
+
+func testDir(directory string, title string) (<-chan runtime.TestResult, error) {
+	files, err := ioutil.ReadDir(directory)
+	if err != nil {
+		return nil, fmt.Errorf("Error " + err.Error())
+	}
+
+	results := make(chan runtime.TestResult)
```
Comment (SimonBaeumer) [inline pkg/app/test_command.go:61]: Using the `TestResult` type outside of the `runtime` package is irritating because it is intended to be just the result of the test suite. BUT the major problem with this is that the output package relies on the result channel and has too many responsiblities.

The output package does:
 - printing results immediately after receiving a result from the channel, see cli.go:47
 - aggregating results
 - print a summary  
 - measuring the duration

If we would take another approach - the `runtime` does not return a `channel` instead we add the output to the runtime struct and use it directly for the output.
The return value of the `runtime.Start` function could be a result which already contains the duration of the execution and some metadata about the whole test run - than the output package could easily print the summary based on the result of our runtime.

If we now want to add the directory feature we could aggregate all test results and get an overall overview of the complete run AND could provide insights into the execution of the each test file.

I hope this was understandable - basically the output package has to many responsibilities which leads to more complex solutions.

**Comment by SimonBaeumer** [comment]
I am really happy that you managed to integrate a feature into my source code and really liked it to review it. 
If you have any questions or did not understood my approach do not hesitate to ask me!

**Comment by dylanhitt** [comment]
@SimonBaeumer Thank you for the feedback. 

Based upon your language I am unsure if you're stating that you plan on implementing what you described above yourself? I would still like to work on this feature with the feedback you provided. Unless you plan on doing it, if so let me know. 

My questions and opinions regarding your feedback are the following: 

- "Removing --file-order flag / synchronous execution". 
Just a question here. Are you stating that the feature being addressed should only be concerned with functioning synchronously for the time being? If so, I assume the app package is still the proper place to begin this execution.
- Unwrap runtime.TestResult
If I understand you here. You want the runtime struct to look like something along these lines:
```go
type Runtime struct {
	Nodes []Node
        Output <output struct/interface>
}
``` 
If so, I love this. Not only for the decoupling but it will make creating a new output formats much easier to conceptually implement. 

- Points to think of:
  - "exiting immediately" - I'm assuming you'd like the out to be displayed until the error appears so if file 57 out of 100 throws err, we will see the output for 1-56 in the log?
  - "What about subdirs?" - Yeah i was looking at walking the dir depending how the feedback went
  - "go.mod" - I think vscode extension I use added those? I'm shocked they weren't removed. I'll look into the vscode extension.

Again thank you for the feedback. Those first few commits were some of the first lines of "real" go I've written. If I can take another stab at this following your guidance would you like me to begin working off a clean branch?

Cheers

**Comment by SimonBaeumer** [comment]
Hey @dylanhitt,
I hope you are doing well :)
I am really glad that you asked this questions and hopefully I can answer them in a clear and structured way which is understandable.

> Based upon your language I am unsure if you're stating that you plan on implementing what you described above yourself? I would still like to work on this feature with the feedback you provided. Unless you plan on doing it, if so let me know.

Yes, I thought about implementing by myself. I am not that good in explaining things in github pull-requests and communicating my ideas in relatively short posts.
If you are interested in implementing it by yourself I could guide you a little bit :)

> "Removing --file-order flag / synchronous execution".
Just a question here. Are you stating that the feature being addressed should only be concerned with functioning synchronously for the time being? If so, I assume the app package is still the proper place to begin this execution.

The `--file-order` flag sorts the output stream by the file name which is misleading because I would assume as a user that it guarantees a synchronous execution of the files.
Further if the `--file-order` flag is set the output stream is blocked. Take a look at `pkg/output/cli.go:55`. 
You append all results to the `testResults` slice until the execution is finished. Therefore all results are printed immediately because inside the `for` loop the printing is skipped.
Try it by yourself with these test files inside a directory and test it with and without the `--file-order` flag:

First file:
```
# Sleep should block the output stream
tests:
  sleep 5:
    exit-code: 0
```
Second file:
```
tests:
  echo hello:
    exit-code: 0
```

I would suggest to 
- sort the directories before execution
- prefer synchronous execution of test suites instead of asynchronous, so basically removing the `goroutine` and `WaitGroup` in the `testDir` function.

> So by write to the runtime struct do you mean something along the lines of storing runtime.TestResult in the runtime struct, or placing the output.OutputWriter in the struct? It sounds like you could mean both or just one? My confusion is that I don't know if you're applying that the runtime package should print the result of each testcase, or would the outputwriter be able to print the individual test results from the metadata provided to it. Would you mind providing an example struct?

I try to re-explain it because my explanation was a "little" bit confusing the last time :D

1. You did not wrap (I miscommunicated this) the `runtime.TestResult`, you created a new result for file errors. File errors imho should terminate the execution directly because syntax errors should never be tolerated. Than there is no need to create a new `runtime.TestResult` for file errors and the special case in the `output/cli.go` package could be removed (making the package simpler).
1. At the moment the `output` package has more responsibilities than formatting and printing our output - additionally it measures the execution duration and aggregates test results, i.e. counting the failure count. 
You could compare it to the `MVC` pattern, our `View` is the `output` package which should not know anything about our business logic - and measuring the exec time is business logic here (because the duration could be used in a `xml` or `json` output too).
Ignore this point at the moment and try to remove the `--file-order` flag and the async execution :)
Honestly I am not sure about this too :/

> Secondly, you state that runtime.Start could just return metadata (such as duration). If we implement the directory feature who gains the responsibility of aggregating? I ask because currently each suite must generate it's own runtime to get the nodes it needs for execution (which is a pattern I like a lot), this means that each file would need it's own runtime, thus generating it's own metadata. I just want to make sure my understanding is clear.

Exactly! Indeed you are doing this already because the `testFile` function does exactly that. It parses the test file, creates a runtime and executes it.

**TL;DR:**

 - [ ] Terminating the test suite if a file error occurs - i.e. `panic` or returning a custom `error`
 - [ ] Removing the file error from the `cli.go` output
 - [ ] Removing `--file-order` flag
 - [ ] Synchronous execution of file tests
 - [ ] Sort files by name
 - [ ] Do not block streaming output in `cli.go`

**Comment by dylanhitt** [comment]
Hey @SimonBaeumer 

Thank you for the thorough response. The propose checklist doesn't seem too bad to implement. Scoping this down will help me greatly! However I am really interested in giving my opinions on "unwrap runtime.TestResult", as I believe it's a tough problem and could lead to some good discussion. Maybe open an issue or we can talk about it another way, I don't know it's up to you. 

Expect some commit soon!

Thank you again

**Comment by SimonBaeumer** [comment]
> Hey @SimonBaeumer
> 
> Thank you for the thorough response. The propose checklist doesn't seem too bad to implement. Scoping this down will help me greatly! However I am really interested in giving my opinions on "unwrap runtime.TestResult", as I believe it's a tough problem and could lead to some good discussion. Maybe open an issue or we can talk about it another way, I don't know it's up to you.
> 
> Expect some commit soon!
> 
> Thank you again

That would be really nice, I opened issue #121 where we can discuss it :)

**Comment by dylanhitt** [comment]
The following should address:

- [x] Terminating the test suite if a file error occurs - i.e. panic or returning a custom error
- [x] Removing the file error from the cli.go output
- [x] Removing --file-order flag - 
- [x] Synchronous execution of file tests - some discussion below
- [x] Sort files by name - `ioutil.ReadDir(directory)` already sorts
- [x] Do not block streaming output in cli.go

@SimonBaeumer I managed to fulfill the requirements however I couldn't remove "the `goroutine` and `WaitGroup` in the `testDir` function", I did change the pattern to loop over `testFile` in the same `goroutine` calling each `testFile` synchronously. This is do to `runtime.Start` returning a different `<-chan TestResult` for each file. There may be a different way to do this as I am new to go and may be overlooking a simple pattern to solve this case. Please show me if so.

But my findings were that we must fan-in all channels into a single one and in order to be able to print to the same `output`. I implemented [testDir](https://github.com/SimonBaeumer/commander/pull/120/files#diff-4014ad36a9981aec07762aacdf6a65efR56) this way as well as implementing. with the [fan-in pattern from here](https://blog.golang.org/pipelines), (i.e I implemented the `merge` function and passed it an `[]chan runtime.TestResult`. I went with the former solution (the current solution), because it seems it would be less of change once we address the shortcomings of the current pattern in #121. I do have some opinions on what could be done to make working with `runtime` easier but I will save them for once this branch is closed. Please let me know if there is a reason why one should prefer `merge` rather than the solution I am proposing I will be happy to change it.

Additional comments:

- We don't need to sort since `ioutil.ReadDir` does it for us 😄 
-  Right now in [test_commad:70](https://github.com/SimonBaeumer/commander/pull/120/files#diff-4014ad36a9981aec07762aacdf6a65efR70) simply skips over subdirs. We can address how we would like to handle valid file types/address subdirs in this PR or in a later one, up to you.
- Question, we support testing against test title when a directory is given? I would prefer no and just return an `err` stating test against title aren't not supported with dirs (or something along those lines).
- Woops forgot to add the `--dir` flag.. Will get that in.

Let me know what you think. I'm beyond thankful for the learning experience from this repo. It makes it very fun to work on.

**Comment by SimonBaeumer** [comment]
You're welcome! 
This is really good work! 
It was my mistake to suggest you just removing the async goroutine, atm it is necessary to run it inside a goroutine like you did :) Very good!

So the last big pain point I see is a feature to execute a single test case. This is controlled by the `title` variable and is triggered if you run commander like `./commander test my-file.yaml my-test`. In this case it only executes the test named `my-test`. 
If commander now scans over a complete directory it checks only the first file if it could not find the test - otherwise it will maybe check multiple test cases. 
Imho this feature can be disabled if a directory should be scanned and just returning an error.

The last ToDos are a lot easier to implement:

- [ ] Removing the `--file-order` flag from the configuration
- [ ] Renaming `input` variable to `testPath`, maybe you find a better name for it
- [ ] Adding `--dir` flag to set explicit directory scanning. If a directory is passed in accidentally it should return an error like `Use --dir to test directories with multiple test files`. 
  - [ ] Maybe the error message could be a little bit clearer, `could not stat file...` or `CreateFile...` is a little bit misleading. It should get resolved by the `--dir` flag too :)
  - [ ] Adding a little help text for the `--dir` flag
  - [ ] Adding that tests can't be filtered with `--dir` enabled
- [ ] Revert changes from `go.sum` to `master`-branch
- [ ] Maybe you could rename my shitty `title` variable to `testFilterTitle`, I was confused by my own variable naming^^
- [ ] `func (w *OutputWriter) printResult(r runtime.TestResult) int` could get rid of the return values and just use a `return` statement in the first `if`-statement

After that we can take a look at the test cases and finish this merge request :) I am excited for your ideas ;)
Thank you for your contributions!

**Comment by dylanhitt** [comment]
I believe that should cover the list.

I didn't quite implement `func (w *OutputWriter) printResult(r runtime.TestResult)` as you might have expected, as dealing with the variable `failed` to decide printing and test failure is a wee bit difficult. I believe the solution for #121 will fix this issue. As you've mentioned earlier passing meta data `struct` to output could be lucrative. 

As for this PR I added an additional `readFile` func to deal with testing if a file is a directory and reading the file into a `[]byte`. This leads me to my next point. Should we start think about creating a struct and a `app.NewTest`(or something similar) function? I say this because as `app` grows having access to struct fields across functions would be useful, but my main idea would be that it makes testing easier. As it will allow us to implement an interface and `moq` the IO operations. Let me know what you think, again could be out of scope.

Final point - documentation. Where would you like it? I assume the readme, but if you have an idea on what section that would be helpful too.

Thank you

**Comment by SimonBaeumer** [comment]
Your code looks good to me!

One last thing I observed:
 - [ ] `pkg/app/test_command.go:74` please use `path.Join` for adding new path elements 
 - Could you update the help text like this? Just added the comment on recursion :) `Execute all test files in a directory sorted by file name, this is not recursive - e.g. /path/to/test_files/`
 - [ ] Updating some comments in the source which I commented
 - [ ] Add `CHANGELOG.md` entry

Documentation:
You can document it inside the `README.md`, `Executing` subsection should be enough. 
If the `--dir` option is available in the `test` command with some help text I would think it is enough.

Refactoring `app`:
I would not do this in this pull request as this is grown a little bit, but feel free to push a little proof-of-concept. Would love to see your ideas.

Output stream:
After thinking about it I wouldn't completely abandon the idea of stream the results to channel but maybe it is better to provide the channel inside the `runtime` on its' instantiation. This would allow a streaming output and aggregating the results. 
We can test it out which approach works better for the output :+1: 

Very nice work! Would love to merge this :)

**Comment by dylanhitt** [comment]
@SimonBaeumer which comments? The only comment in diff that are yours is the function comment from `app.TestCommand`? Maybe I'm just missing something?

And for `CHANGELOG.md` v2.1.0. I assume it's the traditional way of versioning, but just want to make sure. Thanks

**Comment by dylanhitt** [comment]
By the way I almost removed the failed test error highlighting look at `cli.go:70`. Is there a way to test the color of output. I tried adding an integration trying the typical `ANSI escape codes` but couldn't get it working. Maybe another issue to tackle?

**Comment by SimonBaeumer** [comment]
> @SimonBaeumer which comments? The only comment in diff that are yours is the function comment from `app.TestCommand`. Maybe I'm just missing something?
> 
> And for `CHANGELOG.md` v2.1.0. I assume it's the traditional way of versioning, but just want to make sure. Thanks

Ops, I forgot to submit the review comments, sorry about that...
Yes, `v2.1.0` is perfectly fine :)

> By the way I almost removed the failed test error highlighting look at cli.go:70. Is there a way to test the color of output. I tried adding an integration trying the typical ANSI escape codes but couldn't get it working. Maybe another issue to tackle?

If you want to look at it feel free to do so, I would love to see a solution :)
I would not put too much energy into it if you find tasks you are more interested in ;-)

File: commander_linux.yaml [hunk 2]
```diff
@@ -14,4 +14,24 @@ tests:
   test docker:
     command: ./commander test integration/linux/docker.yaml
     stdout: ✓ [docker-host] cat /etc/os-release
-    exit-code: 0
\ No newline at end of file
+    exit-code: 0
+
+  test directory:
```
[the remark this answers is not in the document]
Comment (dylanhitt) [reply commander_linux.yaml:19]: At first it was very confusing, but once i started seeing results form travis-ci it became clearer. I can take a look at the `development.md` some more and see if i can improve it. However once you realize the concept i felt it was very easy to understand.

File: integration/unix/directory_test/place_holder/test [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Hello, I am file in a directory. Don't delete me
```
[the remark this answers is not in the document]
Comment (dylanhitt) [reply integration/unix/directory_test/place_holder/test:1]: Yeah, I actually just realized you can put messages in `.gitkeep`

File: pkg/app/test_command.go [hunk 4]
```diff
@@ -2,32 +2,99 @@ package app
...
+
+	"github.com/SimonBaeumer/commander/pkg/output"
+	"github.com/SimonBaeumer/commander/pkg/runtime"
+	"github.com/SimonBaeumer/commander/pkg/suite"
 )
 
 // TestCommand executes the test argument
-// file is the path to the configuration file
-// title ist the title of test which should be executed, if empty it will execute all tests
+// testPath is the path to the configuration object, an object can be a dir or file
```
[the remark this answers is not in the document]
Comment (dylanhitt) [reply pkg/app/test_command.go:17]: I guess this is the original `vba` in me coming out.. [if you're interested](url). I didn't even notice I switched terms mid sentence.

**Comment by SimonBaeumer** [comment]
@dylanhitt let me know if this pr is ready to review :+1:

**Comment by dylanhitt** [comment]
@SimonBaeumer I believe it is done
